openstack/networking/v2/extensions/vpnaas/services/testing/requests_test.go

```diff
@@ -85,7 +85,71 @@ func TestDelete(t *testing.T) {
 		th.TestHeader(t, r, "X-Auth-Token", fake.TokenID)
 		w.WriteHeader(http.StatusNoContent)
 	})
-
 	res := services.Delete(fake.ServiceClient(), "5c561d9d-eaea-45f6-ae3e-08d1a7080828")
 	th.AssertNoErr(t, res.Err)
 }
+
+func TestUpdate(t *testing.T) {
+	th.SetupHTTP()
+	defer th.TeardownHTTP()
+
+	th.Mux.HandleFunc("/v2.0/vpn/vpnservices/5c561d9d-eaea-45f6-ae3e-08d1a7080828", func(w http.ResponseWriter, r *http.Request) {
+
+		th.TestMethod(t, r, "PUT")
+		th.TestHeader(t, r, "X-Auth-Token", fake.TokenID)
+		th.TestHeader(t, r, "Content-Type", "application/json")
+		th.TestHeader(t, r, "Accept", "application/json")
+		th.TestJSONRequest(t, r, `
+{
+    "vpnservice":{
+        "name": "updatedname",
+        "description": "updated service",
+        "admin_state_up": false
+    }
+}
+      `)
+
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+
+		fmt.Fprintf(w, `
+{
+    "vpnservice": {
+        "router_id": "66e3b16c-8ce5-40fb-bb49-ab6d8dc3f2aa",
+        "status": "PENDING_CREATE",
+        "name": "updatedname",
+        "admin_state_up": false,
+        "subnet_id": null,
+        "tenant_id": "10039663455a446d8ba2cbb058b0f578",
```

Comment:
This is odd. So in this test fixture `project_id` is not here... 

Ignoring the API docs, because they can sometimes have old or incorrect information, what are you seeing in your working environment? You can view the raw API request/response by calling neutron with `neutron --debug`.

Reply:
The API is returning 'project_id' in the response in my working environment when running 
`neutron --debug vpn-service-list`. Should I include it as a field in the result then?

Reply:
Yes please. We always prefer actual API output over API references.

Reply:
Ok. I was under the impression that I should only include the fields reference in the source code [here](https://github.com/openstack/neutron-vpnaas/blob/058469e1b99b647537a5228c6a384d93df5484df/neutron_vpnaas/db/vpn/vpn_models.py).

Reply:
Yeah, sometimes there's a lot of mystery and investigation to this.

You're absolutely right: if the service code is showing fields, that's confident confirmation that the fields are valid.

If the API references are showing fields that the aren't in the code, then there are two possibilities:

1. The API reference is wrong (it happens frequently enough)
2. The field is being added _somewhere_ else in the service code. This happens a lot with Neutron, specifically with `project_id` and `tenant_id`. So the next source of confident information is the API interaction between you and your OpenStack environment. 

For fields like `project_id`, `created_at`, or some other "meta" field, we can be fairly confident that they're _somewhere_ in Neutron and as long as you're seeing them in your API interaction, that's good enough for me. But let's say your API interaction was also showing something like `router_id`. Then we'll really want to dig into the Neutron code (by either grepping the entire code for `router_id` or using the github search box) and figuring out where it's coming from. That's ugly work, but it makes sure Gophercloud is correct.

With regard to test fixtures, please always use fixtures generated yourself rather than fixtures from the API references. We (gophercloud) agreed to enforce this a while back but I occasionally forget to check. Using "real" fixtures helps us make sure the data is correct.

Let me know if all of that makes sense.

Reply:
Yes, that does make a lot of sense. I'll keep that in mind from now on. Thanks for the explanation!